### DISCUSSION TITLE
Use E-Mail in Codeowner on Bitbucket Server platform instead of username

### DISCUSSION THREAD

**Discussion by mikevader** [renovatebot/renovate#30503] [Suggest an Idea]
### Tell us more.

We have a setup with Renovate (38.11.1), Bitbucket Server (8.19.3) and use CodeOwners to add default reviewers.
In Bitbucket we can use E-Mail addresses without issue meaning they are used as default reviewers on new PRs.

We know tried to use this with renovate as well, specifically with the `reviewersFromCodeOwners` flag. This resulted in the following error:
```
DEBUG: CODEOWNERS matched the following files: Dockerfile, values.yaml (repository=RP/renovate-test-repository, branch=renovate/library-alpine-3.x)
DEBUG: CODEOWNERS matched the following users: [{"user":"mike@mycompany.com","score":2}] (repository=RP/renovate-test-repository, branch=renovate/library-alpine-3.x)
DEBUG: Reviewers from code owners: "mike@mycompany.com" (repository=RP/renovate-test-repository, branch=renovate/library-alpine-3.x)
DEBUG: Adding reviewers 'mike@mycompany.com' to #12 (repository=RP/renovate-test-repository, branch=renovate/library-alpine-3.x)
DEBUG: getPr(12) (repository=RP/renovate-test-repository, branch=renovate/library-alpine-3.x)
DEBUG: PUT https://bitbucket.mycompany.com/./rest/api/1.0/projects/RP/repos/renovate-test-repository/pull-requests/12 = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=409 retryCount=0, duration=77) (repository=RP/renovate-test-repository, branch=renovate/library-alpine-3.x)
WARN: Failed to add reviewers (repository=RP/renovate-test-repository, branch=renovate/library-alpine-3.x)
```

After changing the e-mails to the plain usernames it worked without an issue.
But it would still be great to use e-mail addresses as usernames are often quite unreadable (in our case its the employee number)

Would it be possible to add an e-mail to username resolver?

**Comment by rarkins** [comment]
You need to check if Bitbucket Server has an API to resolve email addresses to usernames

**Reply by mikevader** [reply]
I think `/api/latest/admin/users?filter=<filterstring>` could help [API Doc](https://developer.atlassian.com/server/bitbucket/rest/v900/api-group-permission-management/#api-api-latest-admin-users-get)

It returns a list of users which contain the provided string in either usernames, display name or email addresses.

**Reply by viceice** [reply]
looks like you need higher privileges to use that admin API?

**Reply by mikevader** [reply]
Not that I'm aware of. According to the documentation the user must have the permission "LICENSED_USER" which the renovate user should have anyway as this permission is granted to all users which can login.

For most of the POST/PUT/DELETE operations you must be admin.

**Reply by mikevader** [reply]
@rarkins I created a pull request https://github.com/renovatebot/renovate/pull/31254 which should provide a function in this direction.

I tried to test it on a real instance but didn't entirely understand local dev environment
